pkg/hubble/relay/observer.go

```diff
@@ -54,12 +71,55 @@ func (s *Server) GetFlows(req *observerpb.GetFlowsRequest, stream observerpb.Obs
 
 	g, gctx := errgroup.WithContext(ctx)
 	flows := make(chan *observerpb.GetFlowsResponse, qlen)
+	failedPeers := make(chan string, len(peers))
+
+	go func() {
+		g.Wait()
+		close(failedPeers)
+	}()
+
+	go func() {
+		defer close(flows)
```

Comment:
I guess that closing the flows channel here is assumed to be OK even though there are other routines that write into this channel because this goroutine stops only when the context is done, right? I'm not sure it's safe though because if this one returns on `ctx.Done()` before one of the other writers, I think there could be a chance for a tiny race that could lead to a panic (writing to closed channel).

Reply:
You are correct. I have not considered the case where the `ctx.Done()` is racy.

The basic "shutdown" procedure is intended to be as follows (in non-follow-mode):
 - Wait until each each peer in the errgroup finishes its request, this ensures we have collected all `failedPeers`. We indicate this by closing the `failedPeers` channel.
-  The status go-routine sees that it has received all failed peers, therefore it can generate the `RelayStatus` message with the complete list of peers. It then closes `flows`.

In follow-mode, we do not have to option to send back a final status message, since these kind of clients just disconnect. Therefore, we will eventually just get `io.EOF` from `stream.Send` which will cancel the context and tear down all go routines.

But yes, as it is currently we have the problem that we might write to closed channels. I will rework the logic to avoid this.

Reply:
Here is an idea to try and simplify this although I didn't try to implement the logic myself:
What about getting rid of the `failedPeers` channel altogether and leverage your new `RelayStatus` message instead? So when a peer fails, instead of sending `p.Name` in the `failedPeers` chan, you just send a `GetFlowsResponse` with a `RelayStatus` into the flows channel with information about the peer that failed. These status results can then be aggregated by the receiver and sent when appropriate (eg when the timer ticks); it's just a matter of checking for them when reading from the `flows` channel. This way you don't have to create an additional goroutine and can leave the tear-down logic as it was before.